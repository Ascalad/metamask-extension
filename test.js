
const HEX_ADDRESS_REGEX = /^0x[0-9a-fA-F]{40}$/u;
export function isHexAddress(value) {
  return typeof value !== "string" || HEX_ADDRESS_REGEX.test(value);
}

function isHexAddressFast(str) {
  if (typeof str !== "string" || str.length !== 42) return false;
  if (str[0] !== '0' || str[1] !== 'x') return false;
  for (let i = 2; i < 42; i++) {
    let code = str.charCodeAt(i);
    if (!((code >= 48 && code <= 57) || (code >= 97 && code <= 102) || (code >= 65 && code <= 70))) {
      return false;
    }
  }
  return true;
}

function isHexAddressFast2(str) {
  if (typeof str !== "string" || str.length !== 42) return false;
  if (str[0] !== '0' || str[1] !== 'x') return false;
  for (let i = 2; i < 42; i++) {
    let code = str.charCodeAt(i) & 0xDF;
    if (code >= 16 && code <= 25) {
      continue;
    }
    if (code >= 65 && code <= 70) {
      continue;
    }
    return false;
  }
  return true;
}

function isHexAddressFast3(str) {
  if (typeof str !== "string" || str.length !== 42) return false;
  if (str[0] !== '0' || str[1] !== 'x') return false;

  for (let i = 2; i < 42; i++) {
    // Transform the character for case-insensitivity
    const charCode = str.charCodeAt(i) & 0xDF;

    // After transformation, '0'-'9' are in the range 16-25
    const numberRange = (charCode - 16) | (25 - charCode);
    // After transformation, 'a'-'f' are in the range 65-70 ('A'-'F')
    const letterRange = (charCode - 65) | (70 - charCode);

    // This correctly identifies if the charCode is OUTSIDE of both valid ranges.
    // Parentheses are added to fix operator precedence.
    // The condition is now `!= 0` to find invalid characters.
    if (((numberRange & letterRange) & 0x80000000) !== 0) {
      return false;
    }
  }

  return true;
}

// benchmark the two functions
function benchmarkHexAddressFunctions() {
  // console.log(isHexAddressFast3('0x3A400F688DF6ed0DAE7c6c4B17A8f5791Cd8dD87'));
  // return;
  // generate 1000000 test addresses (0x + 40 random cased hex characters)
  const testAddresses = Array.from({ length: 1000000 }, () => {
    // generate an equal split of valid addresses and junk addresses
    // 50% valid addresses
    // 50% junk addresses
    if (Math.random() < 0.5) {
      return `0x${Array.from({ length: 40 }, () => {
        const randomChar = Math.floor(Math.random() * 16).toString(16);
        return Math.random() < 0.5 ? randomChar : randomChar.toUpperCase();
      }).join('')}`;
    }
    // 50% junk addresses, some have wrong length (short of long), wrong prefix,
    // no prefix, not a string, etc
    return Math.random() < 0.5
      ? `0x${Array.from({ length: Math.floor(Math.random() * 40) }, () => {
          const randomChar = Math.floor(Math.random() * 16).toString(16);
          return Math.random() < 0.5 ? randomChar : randomChar.toUpperCase();
        }).join('')}`
      : Math.random() < 0.5
        ? `0x${Array.from({ length: 42 }, () => {
            const randomChar = Math.floor(Math.random() * 16).toString(16);
            return Math.random() < 0.5 ? randomChar : randomChar.toUpperCase();
          }).join('')}`
        : // weird high-order unicode junk
        `0x${Array.from({ length: 40 }, () => {
            const randomChar = Math.floor(Math.random() * 16).toString(16);
            return Math.random() < 0.5 ? randomChar : randomChar.toUpperCase() + String.fromCharCode(Math.floor(Math.random() * 0x10000));
          }).join('')}`
        ;

  });
  console.log(testAddresses[0]);

  console.time('isHexAddress');
  for (let i = 0; i < testAddresses.length; i++) {
    isHexAddress(testAddresses[i]);
  }
  console.timeEnd('isHexAddress');

  console.time('isHexAddressFast');
  for (let i = 0; i < testAddresses.length; i++) {
    isHexAddressFast(testAddresses[i]);
  }
  console.timeEnd('isHexAddressFast');

    console.time('isHexAddressFast2');
  for (let i = 0; i < testAddresses.length; i++) {
    isHexAddressFast2(testAddresses[i]);
  }
  console.timeEnd('isHexAddressFast2');

    console.time('isHexAddressFast3');
  for (let i = 0; i < testAddresses.length; i++) {
    isHexAddressFast3(testAddresses[i]);
  }
  console.timeEnd('isHexAddressFast3');

   for (let i = 0; i < testAddresses.length; i++) {
    if(isHexAddressFast3(testAddresses[i]) !== (isHexAddress(testAddresses[i]))) {
      throw new Error(`Mismatch for address: ${testAddresses[i]}`);
    }
  }
}

benchmarkHexAddressFunctions()