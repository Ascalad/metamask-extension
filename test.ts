/* eslint-disable no-plusplus, no-bitwise */
import { slice } from 'lodash';

type id = number;
type seq = number;
type total = number;
type data = string;

export type ChunkFrame = {
  id: id;
  total: total;
  seq: seq;
  data: data;
};

enum S {
  Id,
  Total,
  Seq,
  Data,
}

function assert(value: boolean, message?: string | number): asserts value {
  if (!value) {
    throw new Error(`Assertion failed: ${message ?? ''}`);
  }
}

/**
 * High-performance raw chunk frame parser.
 *
 * a chunk frame is in the form: `<id>|<total>|<seq>|<data>`, where:
 * - `<id>` is a unique number identifier for the chunked message
 * - `<total>` is the total number of chunks in the message
 * - `<seq>` is the sequence number of this chunk (0-based)
 * - `<data>` is the actual data of this chunk
 *
 * The state machine parser here is about 100x faster than regex+parseInt
 * parsing and  about 10x faster than string split+parseInt parsing.
 *
 * @param x - the value to parse
 * @param input
 * @returns a ChunkFrame if the value is a valid chunk frame, otherwise false
 */
const tryParseChunkFrame = (input: unknown): ChunkFrame | false => {
  if (typeof input !== 'string') {
    return false;
  }
  const { length } = input;
  // shortest legal message is: "0|0|0|" (6 characters)
  if (length < 6) {
    return false;
  }

  let state = S.Id;
  let value = 0;
  let hasDigit = false;

  let id = 0;
  let total = 0;
  let seq = 0;

  for (let i = 0; i < length; ++i) {
    if (state === S.Data) {
      // first byte of data â†’ grab remainder and exit
      return { id, total, seq, data: input.slice(i) };
    }

    const cc = input.charCodeAt(i);
    if (cc >= 48 && cc <= 57) {
      // characters '0'..'9'
      const digit = cc ^ 48; // shift character so "0" is actually `0`
      value = (value << 3) + (value << 1) + digit;
      hasDigit = true;
      continue;
    }
    if (cc === 124) {
      // '|'
      if (!hasDigit) {
        // we didn't find any digits before a delimiter, so this is invalid
        return false;
      }
      switch (state) {
        case S.Id:
          id = value;
          break;
        case S.Total:
          total = value;
          break;
        case S.Seq:
          seq = value;
          break;
      }
      state++;
      value = 0;
      hasDigit = false;
      continue;
    }

    // invalid character
    return false;
  }

  // must have ended in Data state (allows empty data)
  return state === S.Data ? { id, total, seq, data: '' } : false;
};

const tryParseChunkFrame2 = (input: unknown): ChunkFrame | false => {
  // naive approach using indexOf to find pipe positions
  if (typeof input !== 'string') {
    return false;
  }

  // Helper function to check if a string contains only digits
  const isDigitsOnly = (str: string): boolean => {
    if (str.length === 0) {
      return false;
    }
    for (let i = 0; i < str.length; i++) {
      const c = str.charCodeAt(i);
      if (c < 48 || c > 57) {
        return false;
      } // not '0'-'9'
    }
    return true;
  };

  // Find the first 3 pipe positions
  const pipe1 = input.indexOf('|');
  if (pipe1 === -1 || pipe1 === 0) {
    return false;
  } // no pipe or empty id

  const pipe2 = input.indexOf('|', pipe1 + 1);
  if (pipe2 === -1 || pipe2 === pipe1 + 1) {
    return false;
  } // no pipe or empty total

  const pipe3 = input.indexOf('|', pipe2 + 1);
  if (pipe3 === -1 || pipe3 === pipe2 + 1) {
    return false;
  } // no pipe or empty seq

  // Extract the parts directly
  const idStr = input.slice(0, pipe1);
  const totalStr = input.slice(pipe1 + 1, pipe2);
  const seqStr = input.slice(pipe2 + 1, pipe3);
  const data = input.slice(pipe3 + 1);

  // Validate that numeric parts contain only digits
  if (
    !isDigitsOnly(idStr) ||
    !isDigitsOnly(totalStr) ||
    !isDigitsOnly(seqStr)
  ) {
    return false;
  }

  // Parse numbers (now guaranteed to be valid)
  const id = parseInt(idStr, 10);
  const total = parseInt(totalStr, 10);
  const seq = parseInt(seqStr, 10);

  return { id, total, seq, data };
};

const RE = /^(\d+)\|(\d+)\|(\d+)\|(.*)$/u;
const tryParseChunkFrame3 = (input: unknown): ChunkFrame | false => {
  // regex approach - clean and concise
  if (typeof input !== 'string') {
    return false;
  }

  const match = input.match(RE);
  if (!match) {
    return false;
  }

  const [, idStr, totalStr, seqStr, data] = match;

  return {
    id: parseInt(idStr, 10),
    total: parseInt(totalStr, 10),
    seq: parseInt(seqStr, 10),
    data,
  };
};

// benchmark the two functions
const bench = (name: string, fn: () => void) => {
  const start = performance.now();
  for (let i = 0; i < 1_000_00; i++) {
    fn();
  }
  const end = performance.now();
  console.log(`${name}: ${end - start} ms`);
};

const framesToTest = [
  [
    '1234|12345|45|dasfglkijihsdflgkhdsagkfhsdflkghlksdfghklsdfhgklshgklsdfhlkghsdfg',
    {
      id: 1234,
      total: 12345,
      seq: 45,
      data: 'dasfglkijihsdflgkhdsagkfhsdflkghlksdfghklsdfhgklshgklsdfhlkghsdfg',
    },
  ],
  [
    '0|0|0|',
    {
      id: 0,
      total: 0,
      seq: 0,
      data: '',
    },
  ], // minimal
  [
    '1|1|0|data',
    {
      id: 1,
      total: 1,
      seq: 0,
      data: 'data',
    },
  ], // single chunk
  [
    '2|3|1|moredata',
    {
      id: 2,
      total: 3,
      seq: 1,
      data: 'moredata',
    },
  ], // multiple chunks
  [
    '9999|10000|9999|lastchunkdata',
    {
      id: 9999,
      total: 10000,
      seq: 9999,
      data: 'lastchunkdata',
    },
  ], // large numbers
  [
    '1234|5678|90|',
    {
      id: 1234,
      total: 5678,
      seq: 90,
      data: '',
    },
  ], // empty data
  [
    `1234|5678|90|${'a'.repeat(10000)}`,
    {
      // large
      id: 1234,
      total: 5678,
      seq: 90,
      data: 'a'.repeat(10000),
    },
  ],
  [
    `1234|5678|90|${'a'.repeat(100000)}`,
    {
      // very large
      id: 1234,
      total: 5678,
      seq: 90,
      data: 'a'.repeat(100000),
    },
  ],
  [
    '1234|5678|90|data|extra',
    {
      // extra `|`, but this is fine!
      id: 1234,
      total: 5678,
      seq: 90,
      data: 'data|extra',
    },
  ],

  // invalids:
  ['1234|5678|90', false], // missing data
  ['|1234|5678|90|data', false], // missing id
  ['1234|notastring|90|data', false], // non-numeric total
  ['1234|90|0notastring|data', false], // non-numeric seq
];

// check the parsing functions
for (let i = 0; i < framesToTest.length; i++) {
  assert(
    JSON.stringify(tryParseChunkFrame(framesToTest[i][0])) ===
      JSON.stringify(framesToTest[i][1]),
    i,
  );
  assert(
    JSON.stringify(tryParseChunkFrame2(framesToTest[i][0])) ===
      JSON.stringify(framesToTest[i][1]),
    i,
  );
  assert(
    JSON.stringify(tryParseChunkFrame3(framesToTest[i][0])) ===
      JSON.stringify(framesToTest[i][1]),
    i,
  );
}

bench('tryParseChunkFrame (optimized state machine)', () => {
  for (let i = 0; i < framesToTest.length; i++) {
    tryParseChunkFrame(framesToTest[i][0]);
  }
});

bench('tryParseChunkFrame2 (indexOf + validation)', () => {
  for (let i = 0; i < framesToTest.length; i++) {
    tryParseChunkFrame2(framesToTest[i][0]);
  }
});

bench('tryParseChunkFrame3 (regex)', () => {
  for (let i = 0; i < framesToTest.length; i++) {
    tryParseChunkFrame3(framesToTest[i][0]);
  }
});
