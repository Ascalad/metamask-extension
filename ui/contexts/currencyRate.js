import React from 'react';
import PropTypes from 'prop-types';
import { useSelector } from 'react-redux';
import useCurrencyRatePolling from '../hooks/useCurrencyRatePolling';
import { getNativeCurrency } from '../ducks/metamask/metamask';
import useCurrencyRatePolling2 from '../hooks/useCurrencyRatePolling2';
import useTokenRatesPolling from '../hooks/useTokenRatesPolling';

export const CurrencyRateProvider = ({ children }) => {
  const sdf = useSelector(getNativeCurrency);
  // useCurrencyRatePolling();
  useCurrencyRatePolling2();


  // useTokenRatesPolling();

  return <>{children}</>;
};

CurrencyRateProvider.propTypes = {
  children: PropTypes.node,
};
