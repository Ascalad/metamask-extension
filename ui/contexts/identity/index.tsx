import React, { useEffect } from 'react';
import { useAccountSyncing } from '../../hooks/identity/useAccountSyncing';
import {
  useAutoSignIn,
  useAutoSignOut,
} from '../../hooks/identity/useAuthentication';

export const MetamaskIdentityProvider: React.FC = ({ children }) => {
  const { dispatchAccountSyncing, shouldDispatchAccountSyncing } =
    useAccountSyncing();
  const { autoSignIn, shouldAutoSignIn } = useAutoSignIn();
  const { autoSignOut, shouldAutoSignOut } = useAutoSignOut();

  /**
   * Profile syncing effects
   */
  useEffect(() => {
    if (shouldDispatchAccountSyncing) {
      dispatchAccountSyncing();
    }
  }, [shouldDispatchAccountSyncing, dispatchAccountSyncing]);

  /**
   * Authentication effects
   *
   * - Users should be automatically signed in based on various conditions. (see `useAutoSignIn`).
   * - Users should be signed out if basic functionality is disabled. (see `useAutoSignOut`)
   */
  useEffect(() => {
    if (shouldAutoSignIn) {
      // TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31878
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      autoSignIn();
    }
  }, [shouldAutoSignIn, autoSignIn]);

  useEffect(() => {
    if (shouldAutoSignOut) {
      // TODO: Fix in https://github.com/MetaMask/metamask-extension/issues/31878
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      autoSignOut();
    }
  }, [shouldAutoSignOut, autoSignOut]);

  return <>{children}</>;
};
