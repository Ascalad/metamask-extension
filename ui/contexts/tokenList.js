import React from 'react';
import PropTypes from 'prop-types';
import useTokenListPolling from '../hooks/useTokenListPolling';

export const TokenRatesProvider = ({ children }) => {
  useTokenListPolling();

  return <>{children}</>;
};
TokenRatesProvider.propTypes = {
  children: PropTypes.node,
};
