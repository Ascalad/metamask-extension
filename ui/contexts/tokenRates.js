import React from 'react';
import PropTypes from 'prop-types';
import useTokenRatesPolling from '../hooks/useTokenRatesPolling';

export const TokenRatesProvider = ({ children }) => {
  useTokenRatesPolling();

  return <>{children}</>;
};
TokenRatesProvider.propTypes = {
  children: PropTypes.string,
};
