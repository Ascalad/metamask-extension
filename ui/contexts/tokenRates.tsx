import React, { ReactElement } from 'react';
import useTokenRatesPolling from '../hooks/useTokenRatesPolling';

export const TokenRatesProvider = ({
  children,
}: {
  children: ReactElement;
}) => {
  useTokenRatesPolling();

  return <>{children}</>;
};