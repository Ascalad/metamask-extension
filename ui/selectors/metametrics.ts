import { createSelector } from 'reselect';
import { MetaMaskReduxState } from '../store/store';

export const selectFragments = (state: MetaMaskReduxState) =>
  state.metamask.fragments;

export const getDataCollectionForMarketing = (state: MetaMaskReduxState) =>
  state.metamask.dataCollectionForMarketing;

export const getParticipateInMetaMetrics = (state: MetaMaskReduxState) =>
  Boolean(state.metamask.participateInMetaMetrics);

export const getLatestMetricsEventTimestamp = (state: MetaMaskReduxState) =>
  state.metamask.latestNonAnonymousEventTimestamp;

export const selectFragmentBySuccessEvent = createSelector(
  selectFragments,
  (_, fragmentOptions) => fragmentOptions,
  (fragments, fragmentOptions) => {
    if (fragmentOptions.persist) {
      return Object.values(fragments).find(
        (fragment) => fragment.successEvent === fragmentOptions.successEvent,
      );
    }
    return undefined;
  },
);

export const selectFragmentById = createSelector(
  selectFragments,
  (_, fragmentId) => fragmentId,
  (fragments, fragmentId) => {
    // A valid existing fragment must exist in state.
    // If these conditions are not meant we will create a new fragment.
    if (fragmentId && fragments?.[fragmentId]) {
      return fragments[fragmentId];
    }
    return undefined;
  },
);

export const selectMatchingFragment = createSelector(
  (state, params) =>
    selectFragmentBySuccessEvent(state, params.fragmentOptions),
  (state, params) => selectFragmentById(state, params.existingId),
  (matchedBySuccessEvent, matchedById) => matchedById ?? matchedBySuccessEvent,
);
