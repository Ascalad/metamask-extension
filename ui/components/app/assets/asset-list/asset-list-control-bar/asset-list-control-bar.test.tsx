import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { renderWithProvider } from '../../../../../../test/jest';
import { MetaMetricsContext } from '../../../../../contexts/metametrics';
import {
  MetaMetricsEventCategory,
  MetaMetricsEventName,
} from '../../../../../../shared/constants/metametrics';
import AssetListControlBar from './asset-list-control-bar';

describe('AssetListControlBar', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly', () => {
    const store = configureMockStore([thunk])({
      metamask: {
        selectedNetworkClientId: 'selectedNetworkClientId',
        networkConfigurationsByChainId: {
          '0x1': {
            chainId: '0x1',
            defaultRpcEndpointIndex: 0,
            rpcEndpoints: [
              {
                networkClientId: 'selectedNetworkClientId',
              },
            ],
          },
        },
        internalAccounts: {
          selectedAccount: 'selectedAccount',
          accounts: {
            selectedAccount: {},
          },
        },
      },
    });

    const mockTrackEvent = jest.fn();

    const { container } = renderWithProvider(
      <MetaMetricsContext.Provider value={mockTrackEvent}>
        <AssetListControlBar />
      </MetaMetricsContext.Provider>,
      store,
    );

    expect(container).toMatchSnapshot();
  });

  it('should fire metrics event when refresh button is clicked', async () => {
    const store = configureMockStore([thunk])({
      metamask: {
        selectedNetworkClientId: 'selectedNetworkClientId',
        networkConfigurationsByChainId: {
          '0x1': {
            chainId: '0x1',
            defaultRpcEndpointIndex: 0,
            rpcEndpoints: [
              {
                networkClientId: 'selectedNetworkClientId',
              },
            ],
          },
        },
        internalAccounts: {
          selectedAccount: 'selectedAccount',
          accounts: {
            selectedAccount: {},
          },
        },
      },
    });

    const mockTrackEvent = jest.fn();

    const { findByTestId } = renderWithProvider(
      <MetaMetricsContext.Provider value={mockTrackEvent}>
        <AssetListControlBar />
      </MetaMetricsContext.Provider>,
      store,
    );

    const importButton = await findByTestId('import-token-button');
    importButton.click();

    const refreshListItem = await findByTestId('refreshList__button');
    refreshListItem.click();

    expect(mockTrackEvent).toHaveBeenCalledTimes(1);
    expect(mockTrackEvent).toHaveBeenCalledWith({
      category: MetaMetricsEventCategory.Tokens,
      event: MetaMetricsEventName.TokenListRefreshed,
    });
  });

  // it('should re-render when the network filter is changed', async () => {
  //   const { findByTestId } = renderWithProvider(<AssetListControlBar />);

  //   const networkFilter = await findByTestId('network-filter');
  //   networkFilter.click();

  //   // const networkOption = await findByText('Mainnet');
  //   // networkOption.click();

  //   expect(mockTrackEvent).toHaveBeenCalledTimes(1);
  //   expect(mockTrackEvent).toHaveBeenCalledWith({
  //     category: MetaMetricsEventCategory.Tokens,
  //     event: MetaMetricsEventName.TokenListRefreshed,
  //   });
  // });
});
