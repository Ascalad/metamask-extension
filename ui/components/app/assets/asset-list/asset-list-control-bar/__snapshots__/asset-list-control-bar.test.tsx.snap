// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssetListControlBar should render correctly 1`] = `
<div>
  <div
    class="mm-box asset-list-control-bar mm-box--margin-right-2 mm-box--margin-left-2"
  >
    <div
      class="mm-box mm-box--display-flex mm-box--justify-content-flex-end"
    >
      <div
        class="mm-box asset-list-control-bar__buttons mm-box--display-flex mm-box--justify-content-flex-end"
      >
        <div>
          <div
            aria-describedby="tippy-tooltip-1"
            class=""
            data-original-title="Sort by"
            data-tooltipped=""
            style="display: inline;"
            tabindex="0"
          >
            <button
              class="mm-box mm-text mm-button-base mm-button-base--size-sm asset-list-control-bar__button mm-text--body-md-medium mm-box--margin-right-2 mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-default mm-box--rounded-pill"
              data-testid="sort-by-popover-toggle"
            >
              <span
                class="mm-box mm-icon mm-icon--size-sm mm-box--margin-inline-end-0 mm-box--display-inline-block mm-box--color-inherit"
                style="mask-image: url('./images/icons/filter.svg');"
              />
              <span
                class="mm-box mm-text mm-text--inherit mm-box--color-text-default"
              />
            </button>
          </div>
        </div>
        <button
          class="mm-box mm-text mm-button-base mm-button-base--size-sm asset-list-control-bar__button mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-default mm-box--rounded-pill"
          data-testid="import-token-button"
        >
          <span
            class="mm-box mm-icon mm-icon--size-sm mm-box--margin-inline-end-0 mm-box--display-inline-block mm-box--color-inherit"
            style="mask-image: url('./images/icons/more-vertical.svg');"
          />
          <span
            class="mm-box mm-text mm-text--inherit mm-box--color-text-default"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
